### ISSUE TITLE
mockClient .with({ headers: m.anything() }) not mocking if 2 different instances of forge(...) is used

### ISSUE THREAD

**Issue by MaxKanngard** [tulios/mappersmith#316]
# Expected Behavior

When having multiple instances of same `forge` helper function the `mockClient(instance)...when({ headers: m.anything() })` should mock the request.

# Current Behavior

When having multiple instances of same `forge` helper function, then `mockClient(instance)...when({ headers: m.anything() })` is not mocking the request only throwing an error with partial match.

# Previous Behavior 

This worked before in version `2.32.1` when headers was not checked, could be an issue in `isExactMatch` [src/mocks/mock-request.js]

## Failure Logs

  `[Error: [Mappersmith Test] No exact match found for "GET localhost/users" (body: "undefined"; headers: "x-special-header=value"), partial match with "GET localhost/users" (body: "function () return true; }"; headers: "random-header=value"), check your mock definition],`

## Steps to Reproduce

```
import forge from 'mappersmith'
import { mockClient, install, uninstall, m } from 'mappersmith/test'

const createApi = (headers = {}) =>
  forge({
    ignoreGlobalMiddleware: true,
    allowResourceHostOverride: true,
    gatewayConfigs: {
      enableHTTP408OnTimeouts: true,
    },
    middleware: [],
    host: 'localhost',
    resources: {
      User: {
        all: { path: '/users', headers },
      },
    },
  })

describe('mockClient', () => {
  beforeEach(install)
  afterEach(uninstall)

  it('should be able to use function in .with on headers to mock request with 2 different instances', async () => {
    const api = createApi()
    const user = { firstName: 'foo', lastName: 'bar' }
    mockClient(createApi({ 'random-header': 'value' }))
      .resource('User')
      .method('all')
      .with({ body: m.anything(), headers: m.anything() })
      .response([user])
      .assertObject()

    const data = await api.User.all({ headers: { 'X-Special-Header': 'value' } })
    const users = data.data()

    expect(users[0]).toEqual(user)
  })

  it('works if using same headers', async () => {
    const api = createApi()
    const user = { firstName: 'foo', lastName: 'bar' }
    const headers = { 'random-header': 'value' }
    mockClient(createApi(headers))
      .resource('User')
      .method('all')
      .with({ body: m.anything(), headers: m.anything() })
      .response([user])
      .assertObject()

    const data = await api.User.all({ headers })
    const users = data.data()

    expect(users[0]).toEqual(user)
  })

  it('works if using same instance in mockClient', async () => {
    const api = createApi()
    const user = { firstName: 'foo', lastName: 'bar' }
    const headers = { 'random-header': 'value' }
    mockClient(api)
      .resource('User')
      .method('all')
      .with({ body: m.anything(), headers: m.anything() })
      .response([user])
      .assertObject()

    const data = await api.User.all({ headers })
    const users = data.data()

    expect(users[0]).toEqual(user)
  })
})
```

**Comment by Yurickh** [comment]
The actual problem here was a bit more insidious. The fact that you're using two instances is not that relevant, apart from the fact that one of them _has headers by default_ and the other doesn't.
When you use the same instance for both, you're ensuring that both will use the same default headers, so the issue is not present. You can verify that's the case by this line in your failing test example:

```diff
- const api = createApi()
+ const api = createApi({ 'random-header': 'value' })
```

The issue here boiled down to the merging logic inside of a request. We used the spread operator to merge, and because one of the sides of the merging was a function (`m.anything()`), it didn't merge anything, just kept the default headers around.

In a nutshell:
```ts
// different header clients
const api = forge({ ..., headers: { a: 1 } })
const api2  = forge({ ..., headers: { b: 2 }})

// creates a mock for headers { a: 1 }
mockClient(api).with({ headers: m.anything() }) 

// requests with headers { b: 2, c: 3 }, which is not a superset of { a: 1 } so it fails
api2.Make.request({ headers: { c: 3 }})

// same header clients
const api = forge({ ..., headers: { a: 1 })

// creates a mock for headers { a: 1 }
mockClient(api).with({ headers: m.anything() })

// requests with headers { a: 1, c: 3 }, which is a superset of { a: 1 } so it works
api.Make.request({ headers: { c: 3 }})
```

So this is a bug for function matching with default headers. Thanks for reporting! 🙇‍♂️

**Comment by MaxKanngard** [comment]
Ah okay, thanks for explaining.
